# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/ScriptLoader.cpp
# Commit: 1700d5b79273
# Full Hash: 1700d5b79273a6f34d67b7d16470973f0cb65793
# Author: Yulia Startsev <ystartsev@mozilla.com>
# Date: 2022-12-21 21:21:23
# Regressor Bug: 1247687
# File Overlap Count: 1
# Description:
#   Bug 1247687 - Implement csp for Module Workers; r=evilpie,asuth,rpl,ckerschb
#   
#   Depends on D155691
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D156102
# ==============================================================================

diff -r 31888ffde37a -r 1700d5b79273 dom/workers/ScriptLoader.cpp
--- a/dom/workers/ScriptLoader.cpp	Tue Dec 20 20:56:15 2022 +0000
+++ b/dom/workers/ScriptLoader.cpp	Tue Dec 20 20:56:15 2022 +0000
@@ -119,10 +119,9 @@
   nsresult rv;
   nsCOMPtr<nsIURI> uri = aScriptURL;
 
-  // If we have the document, use it. Unfortunately, for dedicated workers
-  // 'parentDoc' ends up being the parent document, which is not the document
-  // that we want to use. So make sure to avoid using 'parentDoc' in that
-  // situation.
+  // Only use the document when its principal matches the principal of the
+  // current request. This means scripts fetched using the Workers' own
+  // principal won't inherit properties of the document, in particular the CSP.
   if (parentDoc && parentDoc->NodePrincipal() != principal) {
     parentDoc = nullptr;
   }
@@ -140,6 +139,7 @@
 
   nsCOMPtr<nsIChannel> channel;
   if (parentDoc) {
+    // This is the path for top level dedicated worker scripts with a document
     rv = NS_NewChannel(getter_AddRefs(channel), uri, parentDoc, aSecFlags,
                        aContentPolicyType,
                        nullptr,  // aPerformanceStorage
@@ -148,6 +148,11 @@
                        aLoadFlags, ios);
     NS_ENSURE_SUCCESS(rv, NS_ERROR_DOM_SECURITY_ERR);
   } else {
+    // This branch is used in the following cases:
+    //    * Shared and ServiceWorkers (who do not have a doc)
+    //    * Static Module Imports
+    //    * ImportScripts
+
     // We must have a loadGroup with a load context for the principal to
     // traverse the channel correctly.
     MOZ_ASSERT(loadGroup);
@@ -161,6 +166,8 @@
     }
 
     if (aClientInfo.isSome()) {
+      // If we have an existing clientInfo (true for all modules and
+      // importScripts), we will use this branch
       rv = NS_NewChannel(getter_AddRefs(channel), uri, principal,
                          aClientInfo.ref(), aController, aSecFlags,
                          aContentPolicyType, aCookieJarSettings,
@@ -538,6 +545,26 @@
   return list;
 }
 
+nsContentPolicyType WorkerScriptLoader::GetContentPolicyType(
+    ScriptLoadRequest* aRequest) {
+  if (aRequest->GetWorkerLoadContext()->IsTopLevel()) {
+    // Implements https://html.spec.whatwg.org/#worker-processing-model
+    // Step 13: Let destination be "sharedworker" if is shared is true, and
+    // "worker" otherwise.
+    return mWorkerRef->Private()->ContentPolicyType();
+  }
+  if (aRequest->IsModuleRequest()) {
+    // Implements the destination for Step 14 in
+    // https://html.spec.whatwg.org/#worker-processing-model
+    //
+    // We need a special subresource type in order to correctly implement
+    // the graph fetch, where the destination is set to "worker" or
+    // "sharedworker".
+    return nsIContentPolicy::TYPE_INTERNAL_WORKER_STATIC_MODULE;
+  }
+  return nsIContentPolicy::TYPE_INTERNAL_WORKER_IMPORT_SCRIPTS;
+}
+
 already_AddRefed<ScriptLoadRequest> WorkerScriptLoader::CreateScriptLoadRequest(
     const nsString& aScriptURL, const mozilla::Encoding* aDocumentEncoding,
     bool aIsMainScript) {
@@ -578,7 +605,7 @@
   } else {
     // Implements part of "To fetch a worklet/module worker script graph"
     // including, setting up the request with a credentials mode,
-    // destination (CSP, TODO).
+    // destination.
 
     // Step 1. Let options be a script fetch options.
     // We currently don't track credentials in our ScriptFetchOptions
@@ -873,10 +900,7 @@
       return rv;
     }
 
-    nsContentPolicyType contentPolicyType =
-        loadContext->IsTopLevel()
-            ? mWorkerRef->Private()->ContentPolicyType()
-            : nsIContentPolicy::TYPE_INTERNAL_WORKER_IMPORT_SCRIPTS;
+    nsContentPolicyType contentPolicyType = GetContentPolicyType(request);
 
     rv = ChannelFromScriptURL(
         principal, parentDoc, mWorkerRef->Private(), loadGroup, ios, secMan,